### DISCUSSION TITLE
Trying to implement an undo-redo feature but running into an odd issue

### DISCUSSION THREAD

**Discussion by Osman-J** [xyflow/xyflow#3364] [Q&A]
I'm trying to implement an undo-redo feature in my react-flow project and am using the useUndoable library to manage the state for both nodes and edges. I'm running into an issue where when I click a node, the onNodesChange callback is fired with a change of "type:position, dragging:false" although I haven't dragged the node.

This causes an issue because when a user clicks a node and does not drag it, it saves that action as a piece of history. So, when a user then clicks "undo" it cycles back to that click and the user has to click "undo" once more to revert back to the piece of history they want.

Here is an example to better explain,

Say a user moves a node from {x: 0, y:0} to {x: 100, y: 100} and then clicks "undo", instead of reverting to {x: 0, y:0} it reverts to that "type:position, dragging:false" change. The user has to click "undo" again and it will finally go back to {x:0, y:0}

As a result when a user simply clicks a node but does not drag it, it saves that as a new piece of history into the useUndoable state.

I suspect this is because the onNodeDragStop event is fired when a node is clicked. I looked through the existing issues and came across #1424 where it seems this behavior was added as part of v11. Is there a way to disable this?

This is my code that updates the state when a node is changed, it pretty much is the boilerplate from the useUndoable react-flow example but modified so it doesn't make a new piece of history every time the position is changed while dragging.

```	
const triggerUpdate = useCallback(
		(t: any, v: any, overwrite: boolean=false) => {
			setElements(e => ({
				nodes: t === 'nodes' ? v : e.nodes,
				edges: t === 'edges' ? v : e.edges,
			}), undefined, overwrite);
		},
		[setElements]
	);
```
```
	const onNodesChange = useCallback(
		(changes: any) => {
      if((changes[0].type === "position" && changes[0].dragging) || (changes[0].type === "select")) {
        triggerUpdate('nodes', applyNodeChanges(changes, elements.nodes), true);
      } else {
        triggerUpdate('nodes', applyNodeChanges(changes, elements.nodes));
      }
		},
		[triggerUpdate, elements.nodes]
	);

```

One idea I had to solve this was to check if the position actually changed when a user clicks a node by saving the position of the node in the onNodeDragStart handler and then comparing it with the end position in the onNodeDragStop handler and saving it to a boolean. Then in the onNodesChange handler check the boolean to see if the node moved or not and only create a new piece of history when it was moved. However, it seems that approach does not work because of the timing when onNodesChange is called, it seems to not have the most updated value of the boolean variable.

Any suggestions on how to fix this or advice on how to better implement this undo/redo feature is appreciated.

Thank you.

**Comment by numankaf** [comment]
I am also facing the same issue, I implemented onNodesChange like following :
```
  const onNodesChange = useCallback(
    (changes: any) => {
      let ignore = ['dimensions', 'position'].includes(changes[0].type);
      if (!changes[0].dragging && changes[0].type === 'position') {
        ignore = false;
      }
      triggerUpdate('nodes', applyNodeChanges(changes, elements.nodes), ignore);
    },
    [triggerUpdate, elements.nodes],
  );

```
As you mentioned,  we have to do undo twice. Did you come up with a solution for this? I would like to hear that.

Thanks in advance
